readme.md
> 🙋 Using YouTube? Check out the original [lite-youtube-embed](https://github.com/paulirish/lite-youtube-embed).  
> 🦾 Need more power? Check out [playerx](https://github.com/playerxo/playerx).

# Lite Vimeo Embed

> #### Renders faster than a sneeze.

Provide videos with a supercharged focus on visual performance.
This custom element renders just like the real thing but approximately 224X faster.

## Comparison

| Normal `<iframe>` Vimeo embed |  `lite-vimeo` |
|---|---|
|  ![Screen Shot 2019-11-03 at 5 23 50 PM](https://user-images.githubusercontent.com/39191/68095560-5c930d00-fe5f-11e9-8104-e73e77a21287.png)   ![Screen Shot 2019-11-03 at 5 21 05 PM](https://user-images.githubusercontent.com/39191/68095562-5d2ba380-fe5f-11e9-8b5f-18f451b0716d.png)  ![Screen Shot 2019-11-03 at 5 19 35 PM](https://user-images.githubusercontent.com/39191/68095565-5d2ba380-fe5f-11e9-835d-85d37df71f52.png)  | ![Screen Shot 2019-11-03 at 5 23 27 PM](https://user-images.githubusercontent.com/39191/68095561-5d2ba380-fe5f-11e9-9393-e2206a64c8bf.png) ![Screen Shot 2019-11-03 at 5 20 55 PM](https://user-images.githubusercontent.com/39191/68095563-5d2ba380-fe5f-11e9-8f9a-f5c4a774cd56.png)  ![Screen Shot 2019-11-03 at 5 20 16 PM](https://user-images.githubusercontent.com/39191/68095564-5d2ba380-fe5f-11e9-908f-7e12eab8b2ad.png) |

## Basic usage

To use the custom embed you will need to:

1. Include the script
1. Use the element `lite-vimeo` markup and scripting
1. Be happy that you're providing a better user experience to your visitors

```html
<!-- Include the custom element script -->
<script type="module" src="https://cdn.jsdelivr.net/npm/lite-vimeo-embed/+esm"></script>

<!-- Use the element. You may define uses before the scripts are parsed and executed. -->
<lite-vimeo videoid="357274789"></lite-vimeo>
```

## Pro-usage

Use this as your HTML, load the script asynchronously, and let the JS progressively enhance it.

```html
<lite-vimeo videoid="357274789" style="background-image: url('https://i.vimeocdn.com/video/810965406.webp?mw=1600&mh=900&q=70');">
  <div class="ltv-playbtn"></div>
</lite-vimeo>
```


CHANGELOG.md
# [0.3.0](https://github.com/luwes/lite-vimeo-embed/compare/v0.2.3...v0.3.0) (2023-11-23)


### Features

* use Vimeo API directly. starting to cost me ([1c033d2](https://github.com/luwes/lite-vimeo-embed/commit/1c033d24f57961da41d02aa95cd037b643e644a5))



## [0.2.3](https://github.com/luwes/lite-vimeo-embed/compare/v0.2.2...v0.2.3) (2023-11-19)


### Bug Fixes

* transition ([7b32cc8](https://github.com/luwes/lite-vimeo-embed/commit/7b32cc86e40d24e0682166940c71f24f699f6ffa))



## [0.2.2](https://github.com/luwes/lite-vimeo-embed/compare/v0.2.1...v0.2.2) (2023-11-19)


### Bug Fixes

* duplicate appends, style fixes ([d05f780](https://github.com/luwes/lite-vimeo-embed/commit/d05f78059fd5a66139cbcfe5c382b0601f73c7e5))
* improve a11y focus ([b694b55](https://github.com/luwes/lite-vimeo-embed/commit/b694b5519a0490de142f0db535832f9eb457141d))



## [0.2.1](https://github.com/luwes/lite-vimeo-embed/compare/v0.2.0...v0.2.1) (2023-11-18)


### Bug Fixes

* Accessibility update ([9239929](https://github.com/luwes/lite-vimeo-embed/commit/9239929eebfd1cf58b5b108a67ea13477d8d6f60))



# [0.2.0](https://github.com/luwes/lite-vimeo-embed/compare/v0.1.0...v0.2.0) (2023-11-18)


### Bug Fixes

* cleanup tooling, keep it simple ([8eb2350](https://github.com/luwes/lite-vimeo-embed/commit/8eb2350149edd0b487bb8f7d8bf065fc05db6043))
* failing thumbnail api ([bb9b27e](https://github.com/luwes/lite-vimeo-embed/commit/bb9b27e22dfdf8b021f52d50af09c5e443886b60))
* remove unneeded code ([66f0e4b](https://github.com/luwes/lite-vimeo-embed/commit/66f0e4b035e9ebbb6ad5b0e9a90470342df22615))



# [0.1.0](https://github.com/luwes/lite-vimeo-embed/compare/v0.0.1...v0.1.0) (2019-12-19)



## [0.0.1](https://github.com/luwes/lite-vimeo-embed/compare/v0.0.1-alpha.0...v0.0.1) (2019-12-16)



## 0.0.1-alpha.0 (2019-12-16)





api/thumb.go
package handler

import (
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"os"
	"path"
	"time"
)

// Pictures internal object provides access to pictures.
type Pictures struct {
	URI         string         `json:"uri,omitempty"`
	Active      bool           `json:"active"`
	Type        string         `json:"type,omitempty"`
	Sizes       []*PictureSize `json:"sizes,omitempty"`
	Link        string         `json:"link,omitempty"`
	ResourceKey string         `json:"resource_key,omitempty"`
}

// PictureSize internal object provides access to picture size.
type PictureSize struct {
	Width              int    `json:"width,omitempty"`
	Height             int    `json:"height,omitempty"`
	Link               string `json:"link,omitempty"`
	LinkWithPlayButton string `json:"link_with_play_button,omitempty"`
}

// Embed internal object provides access to HTML embed code.
type Embed struct {
	HTML string `json:"html,omitempty"`
}

// Video represents a video.
type Video struct {
	URI           string        `json:"uri,omitempty"`
	Name          string        `json:"name,omitempty"`
	Description   string        `json:"description,omitempty"`
	Link          string        `json:"link,omitempty"`
	Duration      int           `json:"duration,omitempty"`
	Width         int           `json:"width,omitempty"`
	Height        int           `json:"height,omitempty"`
	Language      string        `json:"language,omitempty"`
	Embed         *Embed        `json:"embed,omitempty"`
	CreatedTime   time.Time     `json:"created_time,omitempty"`
	ModifiedTime  time.Time     `json:"modified_time,omitempty"`
	ReleaseTime   time.Time     `json:"release_time,omitempty"`
	ContentRating []string      `json:"content_rating,omitempty"`
	License       string        `json:"license,omitempty"`
	Pictures      *Pictures     `json:"pictures,omitempty"`
	Status        string        `json:"status,omitempty"`
	ResourceKey   string        `json:"resource_key,omitempty"`
}

func Handler(w http.ResponseWriter, r *http.Request) {
	rq := r.URL.Query()
    videoid := rq.Get("videoid")
    imageType := rq.Get("type")
    mw := rq.Get("mw")
    mh := rq.Get("mh")
    quality := rq.Get("q")

    if len(videoid) < 1 {
        log.Println("Url Param 'videoid' is missing")
        return
    }

	apiUrl := "https://api.vimeo.com/videos/" + videoid

	spaceClient := http.Client{
		Timeout: time.Second * 2, // Maximum of 2 secs
	}

	req, err := http.NewRequest(http.MethodGet, apiUrl, nil) 
	if err != nil {
		log.Fatal(err)
	}

	q := req.URL.Query()
    q.Add("fields", "pictures")
    req.URL.RawQuery = q.Encode()

	bearer := "Bearer " + os.Getenv("VIMEO_TOKEN")
    // add authorization header to the req
    req.Header.Add("Authorization", bearer)
	req.Header.Set("User-Agent", "lite-vimeo-embed")

	res, getErr := spaceClient.Do(req)
	if getErr != nil {
		log.Fatal(getErr)
	}

	body, readErr := ioutil.ReadAll(res.Body)
	if readErr != nil {
		log.Fatal(readErr)
	}

	video := Video{}
	jsonErr := json.Unmarshal(body, &video)
	if jsonErr != nil {
		log.Fatal(jsonErr)
	}

	imageId := path.Base(video.Pictures.URI)

	imageUrl, err := url.Parse(fmt.Sprintf("https://i.vimeocdn.com/video/%s.%s", imageId, imageType))
	if err != nil {
		log.Fatal(err)
	}

	q = imageUrl.Query()
    q.Add("mw", mw)
    q.Add("mh", mh)
    q.Add("q", quality)
    imageUrl.RawQuery = q.Encode()

    reqImg, imgErr := http.Get(imageUrl.String())
    if imgErr != nil {
        log.Fatal(imgErr)
    }

    // fmt.Fprintf(w, imageUrl.String())

    w.Header().Set("Content-Length", fmt.Sprint(reqImg.ContentLength))
    w.Header().Set("Content-Type", reqImg.Header.Get("Content-Type"))

    _, copyErr := io.Copy(w, reqImg.Body)
    if copyErr != nil {
        log.Fatal(copyErr)
    }

    defer reqImg.Body.Close()
}


src/utils.js

/**
 * Add a <link rel={preload | preconnect} ...> to the head
 */
export function addPrefetch(kind, url, as) {
    const linkElem = document.createElement('link');
    linkElem.rel = kind;
    linkElem.href = url;
    if (as) {
        linkElem.as = as;
    }
    linkElem.crossorigin = true;
    document.head.appendChild(linkElem);
}

export function canUseWebP() {
    var elem = document.createElement('canvas');

    if (elem.getContext && elem.getContext('2d')) {
        // was able or not to get WebP representation
        return elem.toDataURL('image/webp').indexOf('data:image/webp') === 0;
    }

    // very old browser like IE 8, canvas not supported
    return false;
}

/**
 * Get the thumbnail dimensions to use for a given player size.
 *
 * @param {Object} options
 * @param {number} options.width The width of the player
 * @param {number} options.height The height of the player
 * @return {Object} The width and height
 */
export function getThumbnailDimensions({ width, height }) {
    let roundedWidth = width;
    let roundedHeight = height;

    // If the original width is a multiple of 320 then we should
    // not round up. This is to keep the native image dimensions
    // so that they match up with the actual frames from the video.
    //
    // For example 640x360, 960x540, 1280x720, 1920x1080
    //
    // Round up to nearest 100 px to improve cacheability at the
    // CDN. For example, any width between 601 pixels and 699
    // pixels will render the thumbnail at 700 pixels width.
    if (roundedWidth % 320 !== 0) {
        roundedWidth = Math.ceil(width / 100) * 100;
        roundedHeight = Math.round((roundedWidth / width) * height);
    }

    return {
        width: roundedWidth,
        height: roundedHeight
    };
}


rollup.config.js
import resolve from 'rollup-plugin-node-resolve';
import livereload from 'rollup-plugin-livereload';
import babel from 'rollup-plugin-babel';
import {
    terser
} from 'rollup-plugin-terser';
import bundleSize from 'rollup-plugin-size';
import postcss from 'rollup-plugin-postcss';
import copy from 'rollup-plugin-copy';
import cssnano from 'cssnano';

const production = !process.env.ROLLUP_WATCH;

const terserPlugin = terser({
    sourcemap: true,
    warnings: true,
    compress: {
        passes: 2
    },
    mangle: {
        properties: {
            regex: /^_/
        }
    }
});

const config = {
    input: 'src/lite-vimeo-embed.js',
    output: {
        sourcemap: true,
        format: 'iife',
        name: 'ltv',
        file: 'dist/lite-vimeo-embed.js'
    },
    plugins: [
        bundleSize(),
        resolve(),
        postcss({
            extensions: ['.css'],
            plugins: [
                cssnano()
            ]
        }),

        // If we're building for production (npm run build
        // instead of npm run dev), minify
        production && terserPlugin
    ],
    watch: {
        clearScreen: false
    }
};

export default [{
    ...config,
    output: {
        ...config.output,
        file: 'module/lite-vimeo-embed.js',
        format: 'es'
    },
    plugins: [
        ...config.plugins,

        copy({
            targets: [{ src: 'dist', dest: 'public' }]
        }),
    ]
}, {
    ...config,
    output: {
        ...config.output,
        file: 'dist/lite-vimeo-embed.js',
        format: 'umd'
    },
    plugins: [
        ...config.plugins,
        babel(),

        // In dev mode, call `npm run start` once
        // the bundle has been generated
        !production && serve(),

        // Watch the `public` directory and refresh the
        // browser on changes when not in production
        !production && livereload('./public')
    ]
}];

function serve() {
    let started = false;

    return {
        writeBundle() {
            if (!started) {
                started = true;

                require('child_process').spawn('npm', ['run', 'start', '--', '--dev'], {
                    stdio: ['ignore', 'inherit', 'inherit'],
                    shell: true
                });
            }
        }
    };
}


lite-vimeo-embed.js
const style = document.head.appendChild(document.createElement('style'));
style.textContent = /*css*/`

  lite-vimeo {
    aspect-ratio: 16 / 9;
    background-color: #000;
    position: relative;
    display: block;
    contain: content;
    background-position: center center;
    background-size: cover;
    cursor: pointer;
  }

  lite-vimeo > iframe {
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0;
    left: 0;
    border: 0;
  }

  lite-vimeo > .ltv-playbtn {
    font-size: 10px;
    padding: 0;
    width: 6.5em;
    height: 4em;
    background: rgba(23, 35, 34, .75);
    z-index: 1;
    opacity: .8;
    border-radius: .5em;
    transition: opacity .2s ease-out, background .2s ease-out;
    outline: 0;
    border: 0;
    cursor: pointer;
  }

  lite-vimeo:hover > .ltv-playbtn {
    background-color: rgb(0, 173, 239);
    opacity: 1;
  }

  /* play button triangle */
  lite-vimeo > .ltv-playbtn::before {
    content: '';
    border-style: solid;
    border-width: 10px 0 10px 20px;
    border-color: transparent transparent transparent #fff;
  }

  lite-vimeo > .ltv-playbtn,
  lite-vimeo > .ltv-playbtn::before {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate3d(-50%, -50%, 0);
  }

  /* Post-click styles */
  lite-vimeo.ltv-activated {
    cursor: unset;
  }

  lite-vimeo.ltv-activated::before,
  lite-vimeo.ltv-activated > .ltv-playbtn {
    opacity: 0;
    pointer-events: none;
  }
`;

/**
 * Ported from https://github.com/paulirish/lite-youtube-embed
 *
 * A lightweight vimeo embed. Still should feel the same to the user, just MUCH faster to initialize and paint.
 *
 * Thx to these as the inspiration
 *   https://storage.googleapis.com/amp-vs-non-amp/youtube-lazy.html
 *   https://autoplay-youtube-player.glitch.me/
 *
 * Once built it, I also found these:
 *   https://github.com/ampproject/amphtml/blob/master/extensions/amp-youtube (👍👍)
 *   https://github.com/Daugilas/lazyYT
 *   https://github.com/vb/lazyframe
 */
class LiteVimeo extends (globalThis.HTMLElement ?? class {}) {
  /**
   * Begin pre-connecting to warm up the iframe load
   * Since the embed's network requests load within its iframe,
   *   preload/prefetch'ing them outside the iframe will only cause double-downloads.
   * So, the best we can do is warm up a few connections to origins that are in the critical path.
   *
   * Maybe `<link rel=preload as=document>` would work, but it's unsupported: http://crbug.com/593267
   * But TBH, I don't think it'll happen soon with Site Isolation and split caches adding serious complexity.
   */
  static _warmConnections() {
    if (LiteVimeo.preconnected) return;
    LiteVimeo.preconnected = true;

    // The iframe document and most of its subresources come right off player.vimeo.com
    addPrefetch('preconnect', 'https://player.vimeo.com');
    // Images
    addPrefetch('preconnect', 'https://i.vimeocdn.com');
    // Files .js, .css
    addPrefetch('preconnect', 'https://f.vimeocdn.com');
    // Metrics
    addPrefetch('preconnect', 'https://fresnel.vimeocdn.com');
  }

  connectedCallback() {
    this.videoId = this.getAttribute('videoid');

    /**
     * Lo, the vimeo placeholder image!  (aka the thumbnail, poster image, etc)
     * We have to use the Vimeo API.
     */
    let { width, height } = getThumbnailDimensions(this.getBoundingClientRect());
    let devicePixelRatio = window.devicePixelRatio || 1;
    if (devicePixelRatio >= 2) devicePixelRatio *= .75;
    width = Math.round(width * devicePixelRatio);
    height = Math.round(height * devicePixelRatio);

    fetch(`https://vimeo.com/api/v2/video/${this.videoId}.json`)
      .then(response => response.json())
      .then(data => {
        let thumbnailUrl = data[0].thumbnail_large;
        thumbnailUrl = thumbnailUrl.replace(/-d_[\dx]+$/i, `-d_${width}x${height}`);
        this.style.backgroundImage = `url("${thumbnailUrl}")`;
      });

    let playBtnEl = this.querySelector('.ltv-playbtn');
    // A label for the button takes priority over a [playlabel] attribute on the custom-element
    this.playLabel = (playBtnEl && playBtnEl.textContent.trim()) || this.getAttribute('playlabel') || 'Play video';

    if (!playBtnEl) {
      playBtnEl = document.createElement('button');
      playBtnEl.type = 'button';
      playBtnEl.setAttribute('aria-label', this.playLabel);
      playBtnEl.classList.add('ltv-playbtn');
      this.append(playBtnEl);
    }
    playBtnEl.removeAttribute('href');

    // On hover (or tap), warm up the TCP connections we're (likely) about to use.
    this.addEventListener('pointerover', LiteVimeo._warmConnections, {
      once: true
    });

    // Once the user clicks, add the real iframe and drop our play button
    // TODO: In the future we could be like amp-youtube and silently swap in the iframe during idle time
    //   We'd want to only do this for in-viewport or near-viewport ones: https://github.com/ampproject/amphtml/pull/5003
    this.addEventListener('click', this.addIframe);
  }

  addIframe() {
    if (this.classList.contains('ltv-activated')) return;
    this.classList.add('ltv-activated');

    const iframeEl = document.createElement('iframe');
    iframeEl.width = 640;
    iframeEl.height = 360;
    // No encoding necessary as [title] is safe. https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#:~:text=Safe%20HTML%20Attributes%20include
    iframeEl.title = this.playLabel;
    iframeEl.allow = 'accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture';
    // AFAIK, the encoding here isn't necessary for XSS, but we'll do it only because this is a URL
    // https://stackoverflow.com/q/64959723/89484
    iframeEl.src = `https://player.vimeo.com/video/${encodeURIComponent(this.videoId)}?autoplay=1`;
    this.append(iframeEl);

    // Set focus for a11y
    iframeEl.addEventListener('load', iframeEl.focus, { once: true });
  }
}

if (globalThis.customElements && !globalThis.customElements.get('lite-vimeo')) {
  globalThis.customElements.define('lite-vimeo', LiteVimeo);
}

/**
 * Add a <link rel={preload | preconnect} ...> to the head
 */
function addPrefetch(kind, url, as) {
  const linkElem = document.createElement('link');
  linkElem.rel = kind;
  linkElem.href = url;
  if (as) {
    linkElem.as = as;
  }
  linkElem.crossorigin = true;
  document.head.append(linkElem);
}

/**
 * Get the thumbnail dimensions to use for a given player size.
 *
 * @param {Object} options
 * @param {number} options.width The width of the player
 * @param {number} options.height The height of the player
 * @return {Object} The width and height
 */
function getThumbnailDimensions({ width, height }) {
  let roundedWidth = width;
  let roundedHeight = height;

  // If the original width is a multiple of 320 then we should
  // not round up. This is to keep the native image dimensions
  // so that they match up with the actual frames from the video.
  //
  // For example 640x360, 960x540, 1280x720, 1920x1080
  //
  // Round up to nearest 100 px to improve cacheability at the
  // CDN. For example, any width between 601 pixels and 699
  // pixels will render the thumbnail at 700 pixels width.
  if (roundedWidth % 320 !== 0) {
    roundedWidth = Math.ceil(width / 100) * 100;
    roundedHeight = Math.round((roundedWidth / width) * height);
  }

  return {
    width: roundedWidth,
    height: roundedHeight
  };
}
